contracts/connection-router/src/contract/query.rs

```diff
@@ -51,4 +75,54 @@ mod test {
         assert!(result.is_err());
         assert_eq!(result.unwrap_err().current_context(), &Error::ChainNotFound);
     }
+
+    #[test]
+    fn paginated_chains() {
+        let mut deps = mock_dependencies();
+        let chains: Vec<ChainName> = vec![
+            "a-chain".parse().unwrap(),
+            "b-chain".parse().unwrap(),
+            "c-chain".parse().unwrap(),
+            "d-chain".parse().unwrap(),
+        ];
+
+        let endpoints: Vec<ChainEndpoint> = chains
+            .iter()
+            .map(|chain| ChainEndpoint {
+                name: chain.clone(),
+                gateway: Gateway {
+                    address: Addr::unchecked(format!("{} gateway", chain)),
+                },
+                frozen_status: FlagSet::from(GatewayDirection::None),
+            })
+            .collect();
+
+        // save end points to storage
+        for chain_info in endpoints.iter() {
+            assert!(chain_endpoints()
+                .save(deps.as_mut().storage, chain_info.name.clone(), chain_info)
+                .is_ok());
+        }
+
+        // no pagination
+        let result = super::chains(deps.as_ref(), None, None).unwrap();
+        assert_eq!(result.len(), 4);
+        assert_eq!(result, endpoints);
+
+        // with limit
+        let result = super::chains(deps.as_ref(), None, Some(2)).unwrap();
+        assert_eq!(result.len(), 2);
+        assert_eq!(result, vec![endpoints[0].clone(), endpoints[1].clone()]);
+
+        // with page
+        let result =
+            super::chains(deps.as_ref(), Some("c-chain".parse().unwrap()), Some(2)).unwrap();
+        assert_eq!(result.len(), 1);
+        assert_eq!(result, vec![endpoints[3].clone()]);
+
+        // start after last chain
+        let result =
+            super::chains(deps.as_ref(), Some("d-chain".parse().unwrap()), Some(2)).unwrap();
+        assert_eq!(result.len(), 0);
```

Comment:
can you add an "e-chain" test? just to be safe that it works with data that's completely out of scope